dask/array/_array_expr/_blockwise.py

```diff
@@ -0,0 +1,302 @@
+                if i not in chunkss or len(c) > len(chunkss[i]):
+                    chunkss[i] = c
+
+        for k, v in self.new_axes.items():
+            if not isinstance(v, tuple):
+                v = (v,)
+            chunkss[k] = v
+
+        chunks = [chunkss[i] for i in self.out_ind]
+        if self.adjust_chunks:
+            for i, ind in enumerate(self.out_ind):
+                if ind in self.adjust_chunks:
+                    if callable(self.adjust_chunks[ind]):
+                        chunks[i] = tuple(map(self.adjust_chunks[ind], chunks[i]))
+                    elif isinstance(self.adjust_chunks[ind], numbers.Integral):
+                        chunks[i] = tuple(self.adjust_chunks[ind] for _ in chunks[i])
+                    elif isinstance(self.adjust_chunks[ind], (tuple, list)):
+                        if len(self.adjust_chunks[ind]) != len(chunks[i]):
+                            raise ValueError(
+                                f"Dimension {i} has {len(chunks[i])} blocks, adjust_chunks "
+                                f"specified with {len(self.adjust_chunks[ind])} blocks"
+                            )
+                        chunks[i] = tuple(self.adjust_chunks[ind])
+                    else:
+                        raise NotImplementedError(
+                            "adjust_chunks values must be callable, int, or tuple"
+                        )
+            chunks = tuple(chunks)
+        return chunks
+
+    @cached_property
+    def dtype(self):
+        return self.operand("dtype")
+
+    @cached_property
+    def _name(self):
+        if "name" in self._parameters and self.operand("name"):
+            return self.operand("name")
+        else:
+            return (
+                f"{self.token or funcname(self.func).strip('_')}-"
+                + _tokenize_deterministic(
+                    self.func, self.out_ind, self.dtype, *self.args, **self.kwargs
+                )
+            )
+
+    def _layer(self):
+        arginds = [(a, i) for (a, i) in toolz.partition(2, self.args)]
+
+        numblocks = {}
+        dependencies = []
+        arrays = []
+
+        # Normalize arguments
+        argindsstr = []
+
+        for arg, ind in arginds:
+            if ind is None:
+                arg = normalize_arg(arg)
+                arg, collections = _blockwise_unpack_collections_task_spec(arg)
+                dependencies.extend(collections)
+            else:
+                if (
+                    hasattr(arg, "ndim")
+                    and hasattr(ind, "__len__")
+                    and arg.ndim != len(ind)
+                ):
+                    raise ValueError(
+                        "Index string %s does not match array dimension %d"
+                        % (ind, arg.ndim)
+                    )
+                # TODO(expr): This sucks! We should pass the arguments directly into the graph
+                # instead of wrapping them into this stupid class
```

Comment:
let's dial the tone a bit down. I don't think we want this in the code base and we may miss it on final review.

Reply:
Changed the message, but this class **is** annoying